### ISSUE TITLE
Get image tag from Services endpoint

### ISSUE THREAD

**Issue by PatrickEifler** [docker/docker#28484]
I'm currently working on a swarm visualization matrix. For this it would be very convenient if the Service Object would provide the ImageTag and sha256 fields like it does with the Image field. Or at least the ImageID.

I tried different approaches to associate a service with an image to enrich the service object with the above mentioned fields, however no approach is really working. Basically I tried to iterate over the Services -> Tasks -> Containers -> Images to get the accociation done. However the task.Status.ContainerStatus.ContainerID is not equal to any containerIds in the swarm.

Please let me know if there is a solution that I miss.

I think it could be really beneficial to add the fields to each service of the services response in order to make the api even more flexible. At the moment it seems really hard to associate a Service with an Image to extract all the needed information.

**Comment by solidnerd** [comment]
Hey, 
I'm quite not sure if I understand it correctly. But if I can follow your question. Did you checked out the [Docker Swarm Visualizer from Mano Marks](https://github.com/ManoMarks/docker-swarm-visualizer). There is a way to get the ImageTag.

**Comment by PatrickEifler** [comment]
Hi, 
Mano Marks is using the container spec to get the image tag. That is not an issue.
I need to associate an image running in a certain container with the encapsulating service.

**Comment by thaJeztah** [comment]
@PatrickEifler docker 1.12 does not "pin" tasks to a specific sha256. For that reason it's (potentially) possible that tasks on different nodes have a different version of the image.

Docker 1.13 makes improvements in that area, and resolves the sha256 (digest) when deploying/updating a service. This is now stored in the service definition, so that (also in cases where a task is rescheduled), all tasks run the exact same version

For example, docker 1.12;

```
docker service inspect web
[
    {
        "ID": "ai2kvgfqwa8e9i4tz4tpityun",
        "Version": {
            "Index": 171
        },
        "CreatedAt": "2016-11-17T12:03:59.293682444Z",
        "UpdatedAt": "2016-11-17T12:03:59.319964171Z",
        "Spec": {
            "Name": "web",
            "TaskTemplate": {
                "ContainerSpec": {
                    "Image": "nginx-ssl"
                },
                "Resources": {
```

And in docker 1.13

```
docker service inspect web
[
    {
        "ID": "9ioa5wcp8tcmcfkhymkfsstrj",
        "Version": {
            "Index": 11
        },
        "CreatedAt": "2016-11-17T15:26:11.075364571Z",
        "UpdatedAt": "2016-11-17T15:26:11.075364571Z",
        "Spec": {
            "Name": "web",
            "TaskTemplate": {
                "ContainerSpec": {
                    "Image": "nginx:alpine@sha256:5c6be3514a00db611371806c5aeb510dfd285a4b8261c6e5c3323387e396b66b",
                    "DNSConfig": {}
                },
                "Resources": {
```

**Comment by thaJeztah** [comment]
For reference; this is the pull request that added this improvement; https://github.com/docker/docker/pull/28173

There's a change being proposed for the CLI, because the image name including SHA256 is too wide to be useful; https://github.com/docker/docker/pull/28539

**Comment by PatrickEifler** [comment]
Interesting, thanks.

**Comment by nishanttotla** [comment]
@PatrickEifler the current CLI changes are in the PR linked by @thaJeztah. Starting with Docker 1.13, all tasks in a service should be pinned to a specific digest, and the way to retrieve that would be `docker service inspect`. For brevity, other commands like `docker service ls` or `docker service ps` would only show `image:tag`.

**Comment by thaJeztah** [comment]
@PatrickEifler I'll close this issue, because I think this will be addressed by docker 1.13. If you want to give it for a spin, there's a release candidate available (don't use it in production, it's really for testing, and may contain bugs); https://github.com/docker/docker/releases

Feel free to continue the discussion after I closed!
